File: skills/trae-rules-writer/scripts/init_rule.py
#!/usr/bin/env python3
import argparse
import pathlib


def render(template: str, mapping: dict[str, str]) -> str:
    out = template
    for k, v in mapping.items():
        out = out.replace("{{" + k + "}}", v)
    return out


def default_values(name: str, mode: str, description: str, globs: str) -> dict[str, str]:
    always = "true" if mode == "always" else "false"
    if mode == "file" and not globs:
        globs = "*.md"
    if mode in {"intelligent", "manual"}:
        globs = ""
    return {
        "DESCRIPTION": description or f"Rule for {name}",
        "GLOBS": globs,
        "ALWAYS_APPLY": always,
        "RULE_TITLE": name.replace("-", " ").title(),
        "OVERVIEW": "Define concise, actionable behavior constraints for AI.",
        "GUIDELINES": "- Keep rules explicit\n- Use project-relative paths\n- Avoid conflicting instructions",
        "LANG": "markdown",
        "GOOD_EXAMPLE": "# Good\nUse clear, direct guidance.",
        "BAD_EXAMPLE": "# Bad\nUse vague or contradictory guidance.",
        "EXCEPTIONS": "- Keep API fields and file paths unchanged when needed."
    }


def cleanup_frontmatter(text: str) -> str:
    lines = text.splitlines()
    if len(lines) >= 4 and lines[0].strip() == "---":
        if lines[2].strip() == "globs:":
            del lines[2]
    return "\n".join(lines).rstrip() + "\n"


def main() -> int:
    p = argparse.ArgumentParser(prog="init_rule")
    p.add_argument("--skill-dir", required=True)
    p.add_argument("--name", required=True)
    p.add_argument("--mode", choices=["always", "file", "intelligent", "manual"], default="intelligent")
    p.add_argument("--description", default="")
    p.add_argument("--globs", default="")
    p.add_argument("--output-dir", default=".trae/rules")
    args = p.parse_args()

    skill_dir = pathlib.Path(args.skill_dir).resolve()
    template_path = skill_dir / "assets" / "rule.md.template"
    out_dir = pathlib.Path(args.output_dir).resolve()
    out_dir.mkdir(parents=True, exist_ok=True)
    out_file = out_dir / f"{args.name}.md"

    tpl = template_path.read_text(encoding="utf-8")
    mapping = default_values(args.name, args.mode, args.description, args.globs)
    content = render(tpl, mapping)
    content = cleanup_frontmatter(content)
    out_file.write_text(content, encoding="utf-8")
    print(str(out_file))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/trae-rules-writer/init-rule.ts
#!/usr/bin/env node
import * as fs from 'node:fs';
import * as path from 'node:path';

interface Args {
  skillDir?: string;
  name?: string;
  mode: 'always' | 'file' | 'intelligent' | 'manual';
  description: string;
  globs: string;
  outputDir: string;
}

function parseArgs(argv: string[]): Args {
  const args: Args = { mode: 'intelligent', description: '', globs: '', outputDir: '.trae/rules' };
  const required = new Set(['skillDir', 'name']);
  const modeChoices = ['always', 'file', 'intelligent', 'manual'] as const;
  const keyMap: Record<string, keyof Args> = {
    '--skill-dir': 'skillDir',
    '--name': 'name',
    '--mode': 'mode',
    '--description': 'description',
    '--globs': 'globs',
    '--output-dir': 'outputDir',
  };

  for (let i = 0; i < argv.length; i++) {
    const key = keyMap[argv[i]];
    if (key) {
      const val = argv[++i];
      if (val === undefined) {
        console.error(`Missing value for ${argv[i - 1]}`);
        process.exit(1);
      }
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      (args as any)[key] = val;
    }
  }

  for (const r of required) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    if (!(args as any)[r]) {
      console.error(`Missing required argument: --${r.replace(/([A-Z])/g, '-$1').toLowerCase()}`);
      process.exit(1);
    }
  }

  if (!(modeChoices as readonly string[]).includes(args.mode)) {
    console.error(`Invalid mode "${args.mode}". Choose from: ${modeChoices.join(', ')}`);
    process.exit(1);
  }

  return args;
}

function defaultValues(name: string, mode: string, description: string, globs: string): Record<string, string> {
  const alwaysApply = mode === 'always' ? 'true' : 'false';
  if (mode === 'file' && !globs) globs = '*.md';
  if (mode === 'intelligent' || mode === 'manual') globs = '';
  return {
    DESCRIPTION: description || `Rule for ${name}`,
    GLOBS: globs,
    ALWAYS_APPLY: alwaysApply,
    RULE_TITLE: name.replace(/-/g, ' ').replace(/\b\w/g, (c) => c.toUpperCase()),
    OVERVIEW: 'Define concise, actionable behavior constraints for AI.',
    GUIDELINES: '- Keep rules explicit\n- Use project-relative paths\n- Avoid conflicting instructions',
    LANG: 'markdown',
    GOOD_EXAMPLE: '# Good\nUse clear, direct guidance.',
    BAD_EXAMPLE: '# Bad\nUse vague or contradictory guidance.',
    EXCEPTIONS: '- Keep API fields and file paths unchanged when needed.',
  };
}

function render(template: string, mapping: Record<string, string>): string {
  let out = template;
  for (const [k, v] of Object.entries(mapping)) {
    out = out.split(`{{${k}}}`).join(v);
  }
  return out;
}

function cleanupFrontmatter(text: string): string {
  const lines = text.split('\n');
  if (lines.length >= 4 && lines[0].trim() === '---') {
    if (lines[2].trim() === 'globs:') {
      lines.splice(2, 1);
    }
  }
  return lines.join('\n').trimEnd() + '\n';
}

function main(): void {
  const args = parseArgs(process.argv.slice(2));

  const skillDir = path.resolve(args.skillDir!);
  const templatePath = path.join(skillDir, 'assets', 'rule.md.template');
  const outDir = path.resolve(args.outputDir);
  const outFile = path.join(outDir, `${args.name}.md`);

  const tpl = fs.readFileSync(templatePath, 'utf-8');
  const mapping = defaultValues(args.name!, args.mode, args.description, args.globs);
  let content = render(tpl, mapping);
  content = cleanupFrontmatter(content);

  fs.mkdirSync(outDir, { recursive: true });
  fs.writeFileSync(outFile, content, 'utf-8');
  console.log(outFile);
}

main();
